dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/EventTrackerConverterService.java

```diff
@@ -247,21 +248,21 @@ private ProgramStageInstance from( TrackerPreheat preheat, Event event, ProgramS
         programStageInstance.setGeometry( event.getGeometry() );
 
         EventStatus previousStatus = programStageInstance.getStatus();
-        Date completedDate = DateUtils.fromInstant( event.getCompletedAt() );
 
         programStageInstance.setStatus( event.getStatus() );
 
         if ( !Objects.equal( previousStatus, programStageInstance.getStatus() ) && programStageInstance.isCompleted() )
         {
-            programStageInstance.setCompletedDate( completedDate == null ? new Date() : completedDate );
-            programStageInstance
-                .setCompletedBy( event.getCompletedBy() != null ? event.getCompletedBy() : preheat.getUsername() );
+            programStageInstance.setCompletedDate( new Date() );
+            programStageInstance.setCompletedBy( preheat.getUsername() );
         }
 
-        if ( programStage.isEnableUserAssignment() )
+        if ( Boolean.TRUE.equals( programStage.isEnableUserAssignment() ) &&
+            event.getAssignedUser() != null
+            && !event.getAssignedUser().isEmpty() )
         {
-            User assignedUser = preheat.get( User.class, event.getAssignedUser() );
-            programStageInstance.setAssignedUser( assignedUser );
+            Optional<User> assignedUser = preheat.getUserByUsername( event.getAssignedUser().getUsername() );
```

Comment:
can an import also contain a uid instead of a username?

Reply:
Yes, but we are using a preprocessor to always populate the username as well and we have a validator to filter out invalid users.